csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs

```diff
@@ -102,55 +103,145 @@ public static ExitCode Run(string[] args)
-                    return ExitCode.Ok;
+                    logger.LogInfo(" Running tracing analysis.");
+                    return RunTracingAnalysis(analyzerStopwatch, options, logger, canonicalPathCache, pathTransformer);
                 }
 
-                var workingDirectory = Directory.GetCurrentDirectory();
-                var compilerArgs = options.CompilerArguments.ToArray();
-
-                var canonicalPathCache = CanonicalPathCache.Create(logger, 1000);
-                var pathTransformer = new PathTransformer(canonicalPathCache);
-
-                using var analyser = new TracingAnalyser(new LogProgressMonitor(logger), logger, pathTransformer, canonicalPathCache, options.AssemblySensitiveTrap);
+            }
+            catch (Exception ex)  // lgtm[cs/catch-of-all-exceptions]
+            {
+                logger.Log(Severity.Error, "  Unhandled exception: {0}", ex);
+                return ExitCode.Errors;
+            }
+        }
 
-                var compilerArguments = CSharpCommandLineParser.Default.Parse(
-                    compilerVersion.ArgsWithResponse,
-                    workingDirectory,
-                    compilerVersion.FrameworkPath,
-                    compilerVersion.AdditionalReferenceDirectories
-                    );
+        private static ExitCode RunBinaryLogAnalysis(Stopwatch stopwatch, Options options, string binlogPath, ILogger logger, CanonicalPathCache canonicalPathCache, PathTransformer pathTransformer)
+        {
+            logger.LogInfo($"Reading compiler calls from binary log {binlogPath}");
+            try
+            {
+                using var fileStream = new FileStream(binlogPath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                using var reader = BinaryLogReader.Create(fileStream);
 
-                if (compilerArguments is null)
+                // Filter out compiler calls that aren't interesting for examination
+                static bool filter(CompilerCall compilerCall)
                 {
-                    var sb = new StringBuilder();
-                    sb.Append("  Failed to parse command line: ").AppendList(" ", compilerArgs);
-                    logger.Log(Severity.Error, sb.ToString());
-                    ++analyser.CompilationErrors;
-                    return ExitCode.Failed;
+                    return compilerCall.IsCSharp &&
+                        compilerCall.Kind == CompilerCallKind.Regular;
                 }
 
-                if (!analyser.BeginInitialize(compilerVersion.ArgsWithResponse))
+                var allCompilationData = reader.ReadAllCompilationData(filter);
+                var allFailed = true;
+
+                logger.LogInfo($"  Found {allCompilationData.Count} compilations in binary log");
+
+                foreach (var compilationData in allCompilationData)
```

Comment:
- Should we order `allCompilationData` so that we're deterministic?
- Should we run this loop in parallel?

Reply:
Hmm, yes, maybe some ordering under our control would be great; but don't we expect this to be deterministic already? (It looks like `ReadAllCompilerCalls` is not parallel when creating the result result). Have you observed that the order could differ?

Reply:
I haven't experienced any problem with this. So I think it's okay to keep it as it is, but if we see alert wobbles in DCA, this might be one of the reasons.

Reply:
> Should we order allCompilationData so that we're deterministic?

These should be deterministic. If you find that it's not then it's a bug that we'd go and fix in the library. 

This was the first time I'd been directly asked about determinism in this API. Looked through the code and did find one source of non-determinism but it's a fairly straight forward fix. 

https://github.com/jaredpar/complog/issues/139